# Day 5: Binary Boarding

[Brief](https://adventofcode.com/2020/day/5)

- `F` - front
- `B` - back
- `L` - left
- `R` - right
- 128 rows (0 to 127)
- 8 seats per row (0 to 7)

In [89]:
# end will be the number of rows (127) or columns (7)
# lower_char is the character that tells us to keep the lower half of the range
# upper_char is the character that tells us to keep the upper half of the range
# lower is whether the result should be the lower value of the range or higher value
# this is a little bit like a binary search
def binary_space_partitioning(chars, end, lower_char="F", upper_char="B", lower=True):
    start = 0
    
    # go through each of the characters
    for char in chars:
        # calculate the midpoint between the start and end positions
        mid = (start + end) // 2
        if char == lower_char:
            # if the character is equal to the lower_char (e.g. F), keep the lower half
            end = mid
        elif char == upper_char:
            # if the character is equal to the upper_char (e.g. B), keep the higher half
            start = mid + 1
    
    if lower:
        return start
    else:
        return end

def get_seat(chars):
    # calculate the row
    row = binary_space_partitioning(chars[0:7], 127, "F", "B", True)
    col = binary_space_partitioning(chars[7:], 7, "L", "R", False)  
    seat = (row * 8) + col
    
    return (seat, row, col)

In [90]:
assert get_seat("FBFBBFFRLR") == (357, 44, 5)

assert get_seat("BFFFBBFRRR") == (567, 70, 7)
assert get_seat("FFFBBBFRRR") == (119, 14, 7)
assert get_seat("BBFFBBFRLL") == (820, 102, 4)

## Part 1

In [91]:
with open("input.txt", "r") as file:
    boarding_passes = file.read().splitlines()

highest = -1
for boarding_pass in boarding_passes:
    seat, row, col = get_seat(boarding_pass)
    if seat > highest:
        highest = seat

print("The highest seat ID is {}".format(highest))

The highest seat ID is 871


## Part 2

Finding free seats from the list

In [92]:
import numpy as np

all_seats = np.zeros((128, 8))
for boarding_pass in boarding_passes:
    seat, row, col = get_seat(boarding_pass)
    all_seats[row][col] = 1

free_seats = np.argwhere(all_seats == 0)
for seat in free_seats:
    seat_id = (seat[0] * 8) + seat[1]
    print("Free seat ID {} at {}, {}".format(seat_id, seat[0], seat[1]))

Free seat ID 0 at 0, 0
Free seat ID 1 at 0, 1
Free seat ID 2 at 0, 2
Free seat ID 3 at 0, 3
Free seat ID 4 at 0, 4
Free seat ID 5 at 0, 5
Free seat ID 6 at 0, 6
Free seat ID 7 at 0, 7
Free seat ID 8 at 1, 0
Free seat ID 9 at 1, 1
Free seat ID 10 at 1, 2
Free seat ID 11 at 1, 3
Free seat ID 640 at 80, 0
Free seat ID 872 at 109, 0
Free seat ID 873 at 109, 1
Free seat ID 874 at 109, 2
Free seat ID 875 at 109, 3
Free seat ID 876 at 109, 4
Free seat ID 877 at 109, 5
Free seat ID 878 at 109, 6
Free seat ID 879 at 109, 7
Free seat ID 880 at 110, 0
Free seat ID 881 at 110, 1
Free seat ID 882 at 110, 2
Free seat ID 883 at 110, 3
Free seat ID 884 at 110, 4
Free seat ID 885 at 110, 5
Free seat ID 886 at 110, 6
Free seat ID 887 at 110, 7
Free seat ID 888 at 111, 0
Free seat ID 889 at 111, 1
Free seat ID 890 at 111, 2
Free seat ID 891 at 111, 3
Free seat ID 892 at 111, 4
Free seat ID 893 at 111, 5
Free seat ID 894 at 111, 6
Free seat ID 895 at 111, 7
Free seat ID 896 at 112, 0
Free seat ID 897 at 1

The puzzle says that the seat is not at the very front or back, the only seat that is near the middle is **640**!